Cluster-wide local cache invalidation support

kasunbg opened the pull request:
At the moment, there is no way to invalidate the local cache in other cluster nodes. We have to wait till the cache expiry time for that.

This feature provides support for immediate invalidation of the local caches in all the in a cluster when one node removes/modifies a cache entry.

Resolves #1423

kasunbg committed:
Cluster-wide local cache invalidation support

At the moment, there is no way to invalidate the local cache in other cluster nodes. We have to wait till the cache expiry time for that.

This feature provides support for immediate invalidation of the local caches in all the in a cluster when one node removes/modifies a cache entry.

isharak commented on core/javax.cache/src/main/java/org/wso2/carbon/caching/impl/clustering/ClusterCacheInvalidationRequest.java:
```diff
@@ -0,0 +1,93 @@
+package org.wso2.carbon.caching.impl.clustering;
+
```
licenses header is missing

kasunbg replied:
fixed.

kasunbg committed:
adding missing license header

jsdjayanga commented on core/javax.cache/src/main/java/org/wso2/carbon/caching/impl/clustering/ClusterCacheInvalidationRequest.java:
```diff
@@ -0,0 +1,110 @@
+        this.tenantId = tenantId;
+    }
+
+    public ClusterCacheInvalidationRequest() {
```
This public method allows having a dummy ClusterCacheInvalidationRequest without cacheInfo, tenantDomain and tenantId. As there are not setter methods. Hence no way of assigning values. 
Hence I believe this should be private. Please confirm.

kasunbg replied:
This constructor is actually required since this is the Cluster Message that goes to other nodes over the wire. Since the attributes are populated over serialization and de-serialization, we do not need to add setters.

kasunbg committed:
default constructor is not needed for the deserialization process.

kasunbg committed:
pr #1423 - Supporting cache update in addition to cache remove.
Support Cluster-wide local cache invalidation support

At the moment, there is no way to invalidate the local cache in other cluster nodes. We have to wait till the cache expiry time for that.

This feature provides support for immediate invalidation of the local caches in all the in a cluster when one node removes/modifies a cache entry.

afkham commented on core/javax.cache/src/main/java/org/wso2/carbon/caching/impl/CacheImpl.java:
```diff
@@ -124,6 +134,9 @@ public CacheImpl(String cacheName, CacheManager cacheManager) {
                             cacheName, ownerTenantDomain, cacheManager.getName()), new TimestampMapEntryListenerImpl());
         }
+
+        cacheEntryListeners.add(clusterCacheInvalidationReqSender);
```
Why do you need this? When remove is called on the CacheImpl object you can simply do the cache invalidation messaging.

afkham replied:
Ok this is fine. No need to make any changes.

afkham reviewed (commented):
@kasunbg it seems like you have not configured your IDE to automatically add the license header. As Trump would say, BAD!

kasunbg commented:
@afkham as convinced you offline, I do have the IDE setup. :-) This particular issue has occurred when copying class content from the identity-framework repo into the kernel. When copying IDE correctly figured out the package structure, but it didn't add the headers.. :/
